Update Hierarchization

freifrauvonbleifrei opened the pull request:
- removed some code duplication and clutter
- added different mass-conserving hierarchization and dehierarchization operators (currently only for one process for testing purposes, since multiple processes need a multi-recursive communication of point values due to stencilization)
- added the possibility to specify a minimum level for hierarchization, and then removed it again because it actually made the code slower (I believe due to all the level checks). Please do not fast-forward merge this in case a minimum level seems sensible in the future!

obersteiner commented on distributedcombigrid/src/sgpp/distributedcombigrid/fullgrid/DistributedFullGridNonUniform.hpp:
```diff
@@ -708,17 +708,17 @@ class DistributedFullGridNonUniform {
         // to comply with an ordering that mpi understands, we have to reverse
         // our index vectors
```
adapt comment

obersteiner commented on distributedcombigrid/src/sgpp/distributedcombigrid/fullgrid/DistributedFullGridNonUniform.hpp:
```diff
@@ -1217,17 +1217,17 @@ class DistributedFullGridNonUniform {
         // to comply with an ordering that mpi understands, we have to reverse
         // our index vectors
```
adapt comment

obersteiner commented on distributedcombigrid/src/sgpp/distributedcombigrid/fullgrid/FullGrid.hpp:
```diff
@@ -266,7 +266,8 @@ FullGrid<FG_ELEMENT>::FullGrid(DimType dim, LevelType level, bool hasBdrPoints,
   // set the basis function for the full grid
   if (basis == NULL)
-    basis_ = LinearBasisFunction::getDefaultBasis();
+    //TODO deal with memory leak
```
Maybe switch to smart pointers?

freifrauvonbleifrei replied:
Hm, I was considering that (shared_ptr would avoid the ownership ambiguity here). But it would mean wrapping the serialized and sent/received Basis classes into a shared_ptr, and `FullGrid` is rarely ever used in DisCoTec (actually only for tests nowadays). So my suggestion would be to tolerate it for now and to leave the comment in...?

freifrauvonbleifrei commented:
> It is sometimes not directly clear what the methods are doing without analyzing them in detail -> Maybe add some documentation to describe the functions (especially in DistributedHierarchization.hpp). Examples are: what is the direct neighbour? What are the pole neighbours (just the direct neighbours or all of them? Maybe also what is considered to be a "pole" for people that are not used to the term) and so on. This documentation would help readability and makes it easier to spot bugs.

Fair point, thx! I removed the bits of practically unused code, and added some docu and `const` interfaces (at least to the newly-introduced functions). what do you think @obersteiner ?
